pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BucketDelayedDeliveryTracker.java

```diff
@@ -91,9 +98,66 @@ public BucketDelayedDeliveryTracker(PersistentDispatcherMultipleConsumers dispat
         this.sharedBucketPriorityQueue = new TripleLongPriorityQueue();
         this.immutableBuckets = TreeRangeMap.create();
         this.snapshotSegmentLastIndexTable = HashBasedTable.create();
-        this.numberDelayedMessages = 0L;
         ManagedCursor cursor = dispatcher.getCursor();
         this.lastMutableBucket = new MutableBucket(cursor, bucketSnapshotStorage);
+        this.numberDelayedMessages = recoverBucketSnapshot();
+    }
+
+    private long recoverBucketSnapshot() throws RuntimeException {
+        ManagedCursor cursor = this.lastMutableBucket.cursor;
+        cursor.getCursorProperties().keySet().forEach(key -> {
+            if (key.startsWith(DELAYED_BUCKET_KEY_PREFIX)) {
+                String[] keys = key.split(DELIMITER);
+                checkArgument(keys.length == 3);
+                ImmutableBucket immutableBucket =
+                        new ImmutableBucket(cursor, this.lastMutableBucket.bucketSnapshotStorage,
+                                Long.parseLong(keys[1]), Long.parseLong(keys[2]));
+                immutableBuckets.put(Range.closed(immutableBucket.startLedgerId, immutableBucket.endLedgerId),
+                        immutableBucket);
+            }
+        });
+
+        if (immutableBuckets.asMapOfRanges().isEmpty()) {
+            return 0;
+        }
+
+        List<CompletableFuture<Void>> futures = new ArrayList<>(immutableBuckets.asMapOfRanges().size());
+        for (ImmutableBucket immutableBucket : immutableBuckets.asMapOfRanges().values()) {
+            CompletableFuture<Void> future =
+                    immutableBucket.asyncRecoverBucketSnapshotEntry(this::getCutoffTime).thenAccept(indexList -> {
+                        if (CollectionUtils.isEmpty(indexList)) {
+                            return;
+                        }
+                        DelayedIndex lastDelayedIndex = indexList.get(indexList.size() - 1);
+                        synchronized (this.snapshotSegmentLastIndexTable) {
```

Comment:
Is there a race condition for this method? if yes, we'd better make this method thread safe to avoid reading the same entries.